src/Features/Core/Portable/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,403 @@
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var diagnostic = context.Diagnostics.First();
+
+            var root = await document.GetRequiredSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var spanNode = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true)
+                .GetAncestorsOrThis<TExpressionSyntax>().FirstOrDefault();
+            if (spanNode == null)
+                return;
+
+            var hasSolution = TryGetTargetTypeInfo(document,
+                semanticModel, root, diagnostic.Id, spanNode, cancellationToken,
+                out var potentialConversionTypes);
+            if (!hasSolution)
+            {
+                return;
+            }
+
+            if (potentialConversionTypes.Length == 1)
+            {
+                context.RegisterCodeFix(new MyCodeAction(
+                    FeaturesResources.Add_explicit_cast,
+                    c => FixAsync(context.Document, context.Diagnostics.First(), c)),
+                    context.Diagnostics);
+            }
+            else
+            {
+                var actions = ArrayBuilder<CodeAction>.GetInstance();
+
+                // MaximumConversionOptions: we show at most [MaximumConversionOptions] options for this code fixer
+                for (var i = 0; i < Math.Min(MaximumConversionOptions, potentialConversionTypes.Length); i++)
+                {
+                    var targetNode = potentialConversionTypes[i].node;
+                    var conversionType = potentialConversionTypes[i].type;
+                    actions.Add(new MyCodeAction(
+                        GetSubItemName(context, semanticModel, targetNode, conversionType),
+                        _ => ApplySingleConversionToDocumentAsync(document, ApplyFix(root, targetNode, conversionType))));
+                }
+
+                if (potentialConversionTypes.Length > MaximumConversionOptions)
+                {
+                    // If the number of potential conversion types is larger than options we could show, report telemetry
+                    Logger.Log(FunctionId.CodeFixes_AddExplicitCast,
+                        KeyValueLogMessage.Create(m =>
+                        {
+                            m["NumberOfCandidates"] = potentialConversionTypes.Length;
+                        }));
+                }
+
+                context.RegisterCodeFix(new CodeAction.CodeActionWithNestedActions(
+                    FeaturesResources.Add_explicit_cast,
+                    actions.ToImmutableAndFree(), isInlinable: false),
+                    context.Diagnostics);
+            }
+        }
+
+        private static Task<Document> ApplySingleConversionToDocumentAsync(Document document, SyntaxNode currentRoot)
```

Comment:
This isn't needed, just call `document.WithSyntaxRoot` directly

Reply:
In previous PR Cyrus said since it is invoked here https://github.com/dotnet/roslyn/blob/db2dfbd4ecddf8b7f08b48a8c361f8814541c6a4/src/Features/Core/Portable/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs#L126

Format like `_ => Task.FromResult(document.WithSyntaxRoot(ApplyFix(root, targetNode, conversionType))))` is confusing? (not sure if I remember right)